[DRAFT] veo3 blog post (#268)
tigrisdata/tigris-blog: blog/2025-08-26-tigris-veo/index.mdx

this prompt:

```text
Cinematic. An Extreme Close-Up of a scientist’s trembling hands holding a test tube filled with glowing green liquid. Sirens blare in the background as red warning lights flash across the lab. She whispers into a recorder, “If this works, humanity has one last chance."
```

<video controls width="100%">
  <source
    src="/blog/img/blog/veo3-tigris/veo3-scientist.mp4"
    type="video/mp4"
  />
  Download the <a href="/blog/img/blog/veo3-tigris/veo3-scientist.mp4">MP4</a> version.
</video>

Truth be told, the model isn’t perfect (yet), but it’s impressive and a huge
step forward for video generation. Even if the cost of inference remains
expensive (~$3 per generation), it’s still far inexxpesnvie comparatively to
buying and maintaining camera gear. I can see a future where most videos will be
generated with models like Veo3.

In this post, you’ll learn how to build a video generator app using Google’s
Veo3.

## Requirements

- Node: v20.11.0 or later
- [Google AI Studio](https://aistudio.google.com/): prompt enhancement and
  text‑to‑video generation
- [Tigris](https://storage.new): S3‑compatible distributed object storage

## Getting started

Clone the

> [GitHub repository](https://github.com/tigrisdata/examples/tree/00-start/veo3/typescript)

Author: Makes sense

and follow the instructions. You can use the `03-enhance-prompt` branch if you
don’t want to implement the Veo and Tigris code yourself. Project structure:

```text
veo3/
  └─ typescript/
     ├─ public/
     ├─ src/
     │  ├─ app/
     │  │  ├─ api/
     │  │  │  └─ generate/
     │  │  │     └─ route.ts
     │  │  ├─ favicon.ico
     │  │  ├─ globals.css
     │  │  ├─ layout.tsx
     │  │  └─ page.tsx
     ├─ eslint.config.mjs
     ├─ next-env.d.ts
     ├─ next.config.ts
     ├─ package.json
     ├─ postcss.config.mjs
     ├─ README.md
     └─ tsconfig.json
```

## Create a Google AI Studio API key

Let’s wire up Veo3 in our API. First, create a Google API key via Google Cloud
and AI Studio.

<video controls width="100%">
  <source
    src="/blog/img/blog/veo3-tigris/ai-studio-get-api-key.mp4"
    type="video/mp4"
  />
  Download the <a href="/blog/img/blog/veo3-tigris/ai-studio-get-api-key">
    MP4
  </a> version.
</video>

- Create a
  [Google Cloud project](https://developers.google.com/workspace/guides/create-project)

>   (if you don’t have one).

Author: They have a very limited free tier (2 generations).

2. Create an access key with Read/Write. Toggle “Environment variables” and copy
   them into your Next.js `.env`.
3. In “Bucket Rules”, grant Editor permissions.

Add this to `generate/route.ts`:

```typescript
import { NextRequest, NextResponse } from "next/server";
import { GoogleGenAI, GeneratedVideo } from "@google/genai";
import {
  S3Client,
  PutObjectCommand,
  GetObjectCommand,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

const API_KEY = process.env.GOOGLE_API_KEY;
const ai = new GoogleGenAI({
  apiKey: API_KEY,
});
const VEO3_MODEL_NAME = "veo-3.0-generate-preview";

const S3_BUCKET_NAME = "veo3-generations";
const S3_REGION = "auto";
const S3_ENDPOINT = "https://t3.storage.dev";
const S3_FORCE_PATH_STYLE = false;

const s3 = new S3Client({
  region: S3_REGION,
  endpoint: S3_ENDPOINT,
  forcePathStyle: S3_FORCE_PATH_STYLE,
  credentials: {
    accessKeyId: process.env.TIGRIS_STORAGE_ACCESS_KEY_ID!,
    secretAccessKey: process.env.TIGRIS_STORAGE_SECRET_ACCESS_KEY!,
  },
});

async function uploadObject(
  buffer: Buffer,
  contentType: string,
  objectKey: string
) {
  await s3.send(
    new PutObjectCommand({

>       Bucket: S3_BUCKET_NAME,

Reviewer: ```suggestion
      Bucket: TIGRIS_BUCKET_NAME,
```

      Key: objectKey,
      Body: buffer,
      ContentType: contentType,
    })
  );
}

async function getObject(bucket: string, objectKey: string) {
  const command = new GetObjectCommand({
    Bucket: bucket,
    Key: objectKey,
  });
  return await getSignedUrl(s3, command, { expiresIn: 60 * 60 });
}
```

In the code above, we created an S3 client. Tigris is S3 compatible and supports
S3 SDK and CLI. We also added two functions (`uploadObject` and `getObject`) to

> upload the videos to our bucket, and then read the video from it.

Reviewer: ```suggestion
upload the videos to our bucket, and then get videos back from it.
```

Use `uploadObject` and `getObject` in `generateVideoFromText`:

```typescript
async function generateVideoFromText(
  prompt: string,
  numberOfVideos = 1,
): Promise<string[]> {

>   …

Reviewer: Assumption: this is a continuation of existing code in the function. I find it's better to make sure that the code examples are syntactically valid.

```suggestion
  // ...
```

        if (!res.ok) {
          throw new Error(`Failed to fetch video: ${res.status} ${res.statusText}`);
        }
        const contentType = res.headers.get('content-type') || 'application/octet-stream';
        const arrayBuffer = await res.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);

        const extension = contentType.includes('mp4') ? '.mp4' : contentType.includes('quicktime') ? '.mov' : '';
        const uniqueId = (global as any).crypto?.randomUUID?.() || `${Date.now()}-${Math.random().toString(36).slice(2)}`;
        const objectKey = `videos/${uniqueId}${extension || `.mp4`}`;

        await uploadObject(buffer, contentType, objectKey);

        const videoUrl = await getObject(S3_BUCKET_NAME, objectKey);
        return videoUrl;
      }),
    );
  } else {
    throw new Error('No videos generated');
  }
}
```

## Enhance the Veo3 prompt

<video controls width="100%">
  <source
    src="/blog/img/blog/veo3-tigris/enhance-prompt.mp4"
    type="video/mp4"
  />
  Download the <a href="/blog/img/blog/veo3-tigris/enhance-prompt">MP4</a> version.
</video>

With text‑to‑video working and objects stored on Tigris, let’s boost prompt

> quality—and video quality—by structuring the inputs.

Reviewer: ```suggestion
quality and video quality by structuring the inputs.
```

-Wpedantic: the em dash isn't needed here, the intent is that "making the prompt better makes the output better", so the "and video quality" bit flows better inline to the sentence.

  style # @param ["None", "Photorealistic", "Cinematic", "Vintage", "Japanese anime style", "Claymation style", "Stop-motion animation", "In the style of Van Gogh", "Surrealist painting", "Monochromatic black and white", "Vibrant and saturated", "Film noir style", "High-key lighting", "Low-key lighting", "Golden hour glow", "Volumetric lighting", "Backlighting to create a silhouette"]
  temporal_elements # @param ["None", "Slow-motion", "Fast-paced action", "Time-lapse", "Hyperlapse", "Pulsating light", "Rhythmic movement"]

  sound_effects # @param ["None", "Sound of a phone ringing", "Water splashing", "Soft house sounds", "Ticking clock", "City traffic and sirens", "Waves crashing", "Quiet office hum"]
  dialogue # @param {type: 'string'}

Creative rules:
- Prefer concrete, evocative values. Only use "None" when a field is truly irrelevant.
- Keep each value a concise phrase (no multi-sentence essays).
- Be safe and non-offensive.

Output rules:
- Output ONLY a single JSON object. No markdown, no code fences, no commentary.
- Every key MUST appear.

> User idea:\n${rawPrompt}`;

Reviewer: -Wpedantic: Is there a reason you're not using the [structured outputs](https://ai.google.dev/gemini-api/docs/structured-output#javascript) flow here? What you have spelled out will probably work somewhat reliably, but the structured output flow constrains the token generation via BNF grammar such that it's impossible for the model to generate ungrammatical JSON that does not need the schema.

```

You can find the full code here.

Once we have the parameters, synthesize them into a single cinematic prompt:

```typescript
async function enhancePrompt(input: string): Promise<string> {

>   const geminiPrompt = `You are an expert video prompt engineer for Google's Veo model. Your task is to construct the most effective and optimal prompt string using the following keywords. Every single keyword MUST be included. Synthesize them into a single, cohesive, and cinematic instruction. Do not add any new core concepts. Output ONLY the final prompt string, without any introduction or explanation. Mandatory Keywords: ${input}`;

Reviewer: ```suggestion
  const geminiPrompt = `You are an expert cinematic shot planner. Your task is to take the user request and turn it into a description of the shot with details about the framing, setting, scenario, and camera settings. Every single keyword from the user request MUST be included. Synthesize them into a single, cohesive, and cinematic instruction. Do not add any new core concepts. Output ONLY the final prompt string, without any introduction or explanation. Mandatory Keywords: ${input}`;
```

It's better to give the model the goal, not the exact mechanism.

  const result = await ai.models.generateContent({
    model: PROMPT_ENHANCER_MODEL,
    contents: geminiPrompt,
  });

  return result.text || "";
}
```

Finally, use the enhanced prompt in the API route:

```typescript
export async function POST(request: NextRequest) {
  try {
    const body = await request.json();
    const numberOfVideos = 1;

>     const prompt = await enhancePrompt(body.prompt as string | "");

Reviewer: -Wpedantic: you should explode if the user does not specify a prompt.
